File: tools/typescript/src/tools.ts
/**
 * 4 deterministic tools shared by all framework adapters.
 *
 * Pure-TS functions operating on the static datasets in
 * data/candidates.json and data/jobs.json. Mirror semantics of
 * tools/python/tools.py exactly.
 */

import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const DATA_DIR = join(__dirname, "..", "..", "..", "data");

export interface Candidate {
  id: string;
  name: string;
  current_title: string;
  skills: string[];
  years_experience: number;
  location: string;
  remote_ok: boolean;
  bio: string;
  availability: string;
  expected_salary_eur: number;
  previous_roles: { title: string; company: string; years: number }[];
}

export interface Job {
  id: string;
  title: string;
  description: string;
  required_skills: string[];
  nice_to_have_skills: string[];
  min_years_experience: number;
  location: string;
  remote_ok: boolean;
  salary_range_eur: { min: number; max: number };
  contract_type: string;
}

export interface SearchFilters {
  min_years_experience?: number;
  required_skills?: string[];
  location?: string;
  max_salary_eur?: number;
}

export interface MatchScore {
  skill_match_pct: number;
  experience_fit: "match" | "under" | "over";
  location_fit: "match" | "remote_compatible" | "mismatch";
  salary_fit: "in_range" | "below_min" | "above_max";
}

let _candidatesCache: Candidate[] | null = null;
let _jobsCache: Job[] | null = null;
let _candidateById: Map<string, Candidate> | null = null;
let _jobById: Map<string, Job> | null = null;

function getCandidates(): Candidate[] {
  if (_candidatesCache === null) {
    _candidatesCache = JSON.parse(
      readFileSync(join(DATA_DIR, "candidates.json"), "utf-8"),
    ) as Candidate[];
  }
  return _candidatesCache;
}

function getJobs(): Job[] {
  if (_jobsCache === null) {
    _jobsCache = JSON.parse(
      readFileSync(join(DATA_DIR, "jobs.json"), "utf-8"),
    ) as Job[];
  }
  return _jobsCache;
}

function getCandidateById(): Map<string, Candidate> {
  if (_candidateById === null) {
    _candidateById = new Map(getCandidates().map((c) => [c.id, c]));
  }
  return _candidateById;
}

function getJobById(): Map<string, Job> {
  if (_jobById === null) {
    _jobById = new Map(getJobs().map((j) => [j.id, j]));
  }
  return _jobById;
}

/**
 * Search candidates by free-text query and optional filters.
 * Returns up to 10 candidate IDs sorted by match relevance, then id ascending.
 */
export function searchCandidates(query: string, filters?: SearchFilters): string[] {
  const queryTokens = query
    .split(/\s+/)
    .map((t) => t.toLowerCase())
    .filter(Boolean);

  const candidatesWithScore: [number, string][] = [];
  for (const c of getCandidates()) {
    // Filters first (hard reject)
    if (
      filters?.min_years_experience !== undefined &&
      c.years_experience < filters.min_years_experience
    ) {
      continue;
    }
    if (filters?.required_skills) {
      const candSkillsLower = new Set(c.skills.map((s) => s.toLowerCase()));
      if (!filters.required_skills.every((s) => candSkillsLower.has(s.toLowerCase()))) {
        continue;
      }
    }
    if (filters?.location !== undefined) {
      const locLower = filters.location.toLowerCase();
      if (locLower === "remote") {
        if (!c.remote_ok) continue;
      } else if (locLower !== c.location.toLowerCase()) {
        continue;
      }
    }
    if (
      filters?.max_salary_eur !== undefined &&
      c.expected_salary_eur > filters.max_salary_eur
    ) {
      continue;
    }

    const haystack = (
      c.current_title +
      " " +
      c.skills.join(" ") +
      " " +
      c.bio
    ).toLowerCase();

    let score = 0;
    for (const t of queryTokens) {
      if (haystack.includes(t)) score += 1;
    }

    if (score > 0 || queryTokens.length === 0) {
      candidatesWithScore.push([score, c.id]);
    }
  }

  candidatesWithScore.sort((a, b) => {
    if (b[0] !== a[0]) return b[0] - a[0];
    return a[1].localeCompare(b[1]);
  });

  return candidatesWithScore.slice(0, 10).map(([, id]) => id);
}

/**
 * Get full profile for a candidate, or null if not found.
 */
export function getCandidateProfile(candidateId: string): Candidate | null {
  return getCandidateById().get(candidateId) ?? null;
}

/**
 * Score how well a candidate matches a job. Returns breakdown, not aggregate.
 */
export function scoreMatch(
  candidateId: string,
  jobId: string,
): MatchScore | { error: string } {
  const cand = getCandidateById().get(candidateId);
  const job = getJobById().get(jobId);
  if (!cand || !job) {
    return { error: `unknown candidate_id=${candidateId} or job_id=${jobId}` };
  }

  // Skill match
  const candSkillsLower = new Set(cand.skills.map((s) => s.toLowerCase()));
  const requiredLower = job.required_skills.map((s) => s.toLowerCase());
  const matchedRequired = requiredLower.filter((s) => candSkillsLower.has(s)).length;
  const skill_match_pct =
    requiredLower.length > 0
      ? Math.round((100 * matchedRequired) / requiredLower.length)
      : 0;

  // Experience fit
  let experience_fit: MatchScore["experience_fit"];
  if (cand.years_experience >= job.min_years_experience) {
    const yearsAbove = cand.years_experience - job.min_years_experience;
    experience_fit = yearsAbove >= 5 ? "over" : "match";
  } else {
    experience_fit = "under";
  }

  // Location fit
  let location_fit: MatchScore["location_fit"];
  if (cand.location.toLowerCase() === job.location.toLowerCase()) {
    location_fit = "match";
  } else if (job.remote_ok && cand.remote_ok) {
    location_fit = "remote_compatible";
  } else {
    location_fit = "mismatch";
  }

  // Salary fit
  const sal = cand.expected_salary_eur;
  let salary_fit: MatchScore["salary_fit"];
  if (sal < job.salary_range_eur.min) {
    salary_fit = "below_min";
  } else if (sal > job.salary_range_eur.max) {
    salary_fit = "above_max";
  } else {
    salary_fit = "in_range";
  }

  return { skill_match_pct, experience_fit, location_fit, salary_fit };
}

/**
 * List all jobs with id/title/location only (lightweight).
 */
export function listJobs(): { id: string; title: string; location: string }[] {
  return getJobs().map((j) => ({ id: j.id, title: j.title, location: j.location }));
}

File: tools/python/tools.py
"""4 deterministic tools shared by all framework adapters.

These are pure-Python functions that operate on the static datasets in
`data/candidates.json` and `data/jobs.json`. No LLM calls inside tools.
"""

from __future__ import annotations

import json
from functools import cache
from pathlib import Path
from typing import Any

DATA_DIR = Path(__file__).resolve().parents[2] / "data"


@cache
def _candidates() -> list[dict[str, Any]]:
    return json.loads((DATA_DIR / "candidates.json").read_text())


@cache
def _jobs() -> list[dict[str, Any]]:
    return json.loads((DATA_DIR / "jobs.json").read_text())


@cache
def _candidate_by_id() -> dict[str, dict[str, Any]]:
    return {c["id"]: c for c in _candidates()}


@cache
def _job_by_id() -> dict[str, dict[str, Any]]:
    return {j["id"]: j for j in _jobs()}


def search_candidates(
    query: str,
    filters: dict[str, Any] | None = None,
) -> list[str]:
    """Search candidates by free-text query and optional filters.

    Args:
        query: Free-text query, matched (case-insensitive, OR on tokens) against
            current_title, skills, and bio.
        filters: Optional dict with keys:
            - min_years_experience: int
            - required_skills: list[str] (AND match)
            - location: str (exact match or "remote" — matches remote_ok=True)
            - max_salary_eur: int

    Returns:
        List of candidate IDs (max 10), sorted by number of matched query tokens
        descending then by id ascending for determinism.
    """
    filters = filters or {}
    query_tokens = [t.lower() for t in query.split() if t]

    candidates_with_score: list[tuple[int, str]] = []
    for c in _candidates():
        # Filters first (hard reject)
        if (min_y := filters.get("min_years_experience")) is not None:
            if c["years_experience"] < min_y:
                continue
        if req_skills := filters.get("required_skills"):
            cand_skills_lower = {s.lower() for s in c["skills"]}
            if not all(s.lower() in cand_skills_lower for s in req_skills):
                continue
        if (loc := filters.get("location")) is not None:
            loc_lower = loc.lower()
            if loc_lower == "remote":
                if not c["remote_ok"]:
                    continue
            elif loc_lower != c["location"].lower():
                continue
        if (max_sal := filters.get("max_salary_eur")) is not None:
            if c["expected_salary_eur"] > max_sal:
                continue

        # Score by token matches
        haystack = (
            c["current_title"].lower()
            + " "
            + " ".join(s.lower() for s in c["skills"])
            + " "
            + c["bio"].lower()
        )
        score = sum(1 for t in query_tokens if t in haystack)
        if score > 0 or not query_tokens:
            candidates_with_score.append((score, c["id"]))

    candidates_with_score.sort(key=lambda x: (-x[0], x[1]))
    return [cid for _, cid in candidates_with_score[:10]]


def get_candidate_profile(candidate_id: str) -> dict[str, Any] | None:
    """Get full profile for a candidate, or None if not found."""
    return _candidate_by_id().get(candidate_id)


def score_match(candidate_id: str, job_id: str) -> dict[str, Any]:
    """Score how well a candidate matches a job.

    Returns the breakdown (skill_match_pct, experience_fit, location_fit,
    salary_fit) — NOT an aggregate score. The agent must synthesize.
    """
    cand = _candidate_by_id().get(candidate_id)
    job = _job_by_id().get(job_id)
    if cand is None or job is None:
        return {"error": f"unknown candidate_id={candidate_id} or job_id={job_id}"}

    # Skill match: % of job required_skills present in candidate skills
    cand_skills_lower = {s.lower() for s in cand["skills"]}
    required_lower = [s.lower() for s in job["required_skills"]]
    matched_required = sum(1 for s in required_lower if s in cand_skills_lower)
    skill_match_pct = (
        round(100 * matched_required / len(required_lower)) if required_lower else 0
    )

    # Experience fit
    if cand["years_experience"] >= job["min_years_experience"]:
        years_above = cand["years_experience"] - job["min_years_experience"]
        # 5+ years above min for a junior role = "over"
        experience_fit = "over" if years_above >= 5 else "match"
    else:
        experience_fit = "under"

    # Location fit
    if cand["location"].lower() == job["location"].lower():
        location_fit = "match"
    elif job["remote_ok"] and cand["remote_ok"]:
        location_fit = "remote_compatible"
    else:
        location_fit = "mismatch"

    # Salary fit
    sal = cand["expected_salary_eur"]
    smin, smax = job["salary_range_eur"]["min"], job["salary_range_eur"]["max"]
    if sal < smin:
        salary_fit = "below_min"
    elif sal > smax:
        salary_fit = "above_max"
    else:
        salary_fit = "in_range"

    return {
        "skill_match_pct": skill_match_pct,
        "experience_fit": experience_fit,
        "location_fit": location_fit,
        "salary_fit": salary_fit,
    }


def list_jobs() -> list[dict[str, Any]]:
    """List all jobs with id/title/location only (lightweight)."""
    return [
        {"id": j["id"], "title": j["title"], "location": j["location"]}
        for j in _jobs()
    ]
